### DISCUSSION TITLE
ini-file: development defines `disable_functions`

### DISCUSSION THREAD

**Discussion by mfn** [shivammathur/setup-php#573] [Q&A / Help] [answered]
Hi,

following https://github.com/shivammathur/setup-php/issues/450#issuecomment-1032694715 I already use `ini-file: development` for some time.

Today I was about to run some code making use of `pcntl_async_signals` and got this:
```
ErrorException: pcntl_async_signals() has been disabled for security reasons
```

After some debugging I figured it's related to `ini-file: development`, `ini-file: production` does not have it, see
- [`ini-file: development` workflow example](https://github.com/mfn/setup-php-test/runs/5446016243?check_suite_focus=true#step:3:207)
- [no ini-file specified workflow example](https://github.com/mfn/setup-php-test/runs/5446035658?check_suite_focus=true#step:3:207)

I then checked where this is coming from
- [the official php-src does not list any functions](https://github.com/php/php-src/blob/ea62b8089acef6551d6cece5dfb2ce0b040a7b83/php.ini-development#L323)
  I also checked older PHP version (like: [7.4 as in my case, same empty value](https://github.com/php/php-src/blob/ed949a197333ee0fef5a4f6b497e8e39158730c4/php.ini-development#L312))
- I checked a random VM with Ubuntu 18.04 (I know, old) lying around and then saw this in there:
  ```
  vagrant@dev /etc/php/7.4 $ grep -ri disable_functions *
  cli/php.ini:disable_functions =
  fpm/pool.d/www.conf:; extension_dir. Defining 'disable_functions' or 'disable_classes' will not
  fpm/php.ini:disable_functions = pcntl_alarm,pcntl_fork,pcntl_waitpid,pcntl_wait,pcntl_wifexited,pcntl_wifstopped,pcntl_wifsignaled,pcntl_wifcontinued,pcntl_wexitstatus,pcntl_wtermsig,pcntl_wstopsig,pcntl_signal,pcntl_signal_get_handler,pcntl_signal_dispatch,pcntl_get_last_error,pcntl_strerror,pcntl_sigprocmask,pcntl_sigwaitinfo,pcntl_sigtimedwait,pcntl_exec,pcntl_getpriority,pcntl_setpriority,pcntl_async_signals,pcntl_unshare,
  ```
  But this is php-fpm, not cli.

Maybe my expectations are wrong, but I would not have expected for the CI tests that `disable_functions` has a value. I could not find a reference in my `@verbose` workflows that fpm is used in some way.

Any ideas?

(note: workaround for me now is easy: I just explicitly set `ini-value` with an empty `disable_functions`, this works)

Thanks!

**Comment by shivammathur** [comment, marked as answer]
This is from ondrej's builds. It comes with 3 ini files.

- `php.ini-development`: Used for all SAPI, has `disabled_functions` set.
- `php.ini-production`: Used for all SAPI expect `cli`, this also has `disabled_functions` set.
- `php.ini-production.cli`: Used for `cli`, does not have `disabled_functions` set.

So the differentiation you mentioned between `cli` and `fpm` is in the case of `production` php.ini.

For 7.4, these will be in `/usr/lib/php/7.4`, and are used when a SAPI is configured.

Source: https://salsa.debian.org/php-team/php/-/blob/debian/main/7.4/debian/rules#L504

As you mentioned, the workaround is trivial, so I would not like to change this.

**Reply by susnux** [reply]
I run into the same problem today: `ini-file: development` or `ini-file: none` both set the `disabled_functions` property.

For everybody looking for the solution: While setting `ini-values: disable_functions=,` does **not** work, you **should** set `ini-values: disable_functions=""` to get it work.

(It is really confusing that the php.ini does not match the one from the php docs).
